### DISCUSSION TITLE
Asking about SDP Constraints format

### DISCUSSION THREAD

**Discussion by mrfikri-ai** [cvxpy/cvxpy#2472] [Q&A] [answered]
Hi cvxpy folks,

I just new to learn this library especially related to SDP.
The problem that I have is integer QP and want to relax it using SDP.
I set some parameters like X - xx^T >> 0, I think the matrix form will be [[1, x.T],[x, X]] >>0
Below is the code that I developed. 

`def sdp_qp(values, cost, Q, p, r):
    agent_num = values.shape[0]
    task_num = len(cost)

    # Define for constraints
    A = np.kron(np.eye(len(values)), np.ones((1, len(cost))))
    C = np.kron(np.ones((1, len(values))), np.eye(len(cost)))
    b = np.ones((len(values), 1))
    d = np.ones((len(cost), 1))

    X = cp.Variable((agent_num * task_num, agent_num * task_num), symmetric=True)
    x = cp.Variable(agent_num * task_num)
    W = cp.Variable((agent_num * task_num + 1, agent_num * task_num + 1), PSD=True)

    constraints = [
        A @ x == b.flatten(),  # To make the shape compatible to l.h.s
        -C @ x <= -d.flatten(),  # To make the shape compatible to l.h.s
        W[0, 0] == 1,          # top-left of W is 1
        W[1:, 0] == x,         # first column of W (excluding the top element) is x
        W[0, 1:] == x.T,       # first row of W (excluding the top element) is x^T
        W[1:, 1:] == X,        # bottom-right of W is X
        cp.diag(X) == x        # diag(X) = x
    ]
    obj = cp.Problem(cp.Minimize(cp.trace(Q @ X) + p @ x + r), constraints)
    start_time = time.time()
    obj.solve()
    end_time = time.time() - start_time
    xval = (x.value)
    return xval, end_time`

I am wondering if the constraints [[1, x.T],[x, X]] >>0 that I defined as W is correct.
Because when I tried to use cp.bmat
`expr= cp.bmat([
[cp.Constant(1), x.T.reshape((1, task_num * agent_num], [x.reshape((task_num * agent_num)), X]
])` 
I got an error 'Variable' does not have reshape ...

Could you give me a suggestion?

**Comment by Transurgeon** [comment, marked as answer]
Hello @mrfikri-ai sorry for the late reply. 
Has this issue been resolved?
it seems like you might be missing a closing parentheses here?
```py
x.T.reshape((1, task_num * agent_num]
```
should it be like the following?
```py
x.T.reshape((1, task_num * agent_num)]
```

**Reply by mrfikri-ai** [reply]
Hi,

My bad I didn't realize to close this question.
Yes, I found the error because the missing parentheses
Thank you!
